### ISSUE TITLE
Loss Problem

### ISSUE THREAD

**Issue by Morayabb** [hanlinwu/SADN#6]
Thank you very much for your excellent work. When the code is reproduced, it will appear: loss=nan, how should I solve it? I hope you can help solve it in your spare time. Thanks again！！

**Comment by hanlinwu** [comment]
Thank you for your attention to our work, I did not reproduce the "loss=nan" problem. Please make sure you run the code using pytorch==1.10.0, pytorch lightning==1.5.5. If you have any further questions, feel free to contact me.

**Comment by Morayabb** [comment]
Thank you very much for your reply! I run this code in pytorch=1.10.0, pytorch-lighting==1.5.5. So I would like to ask, what is the Python version?

**Comment by hanlinwu** [comment]
The python version is 3.8.10 (I don't think this is the reason), please try downgrading the Pillow version to 9.5.0.

Here are the versions of all packages:
Package                        Version
------------------------------ --------------------
absl-py                        1.0.0
aiohttp                        3.9.5
aiosignal                      1.3.1
anyio                          3.4.0
argon2-cffi                    21.1.0
async-timeout                  4.0.3
attrs                          21.2.0
Babel                          2.9.1
backcall                       0.2.0
bleach                         4.1.0
brotlipy                       0.7.0
cachetools                     4.2.4
certifi                        2021.5.30
cffi                           1.14.6
chardet                        4.0.0
conda                          4.10.3
conda-package-handling         1.7.3
cryptography                   3.4.7
cycler                         0.11.0
debugpy                        1.5.1
decorator                      5.1.0
defusedxml                     0.7.1
easydict                       1.13
entrypoints                    0.3
fonttools                      4.28.2
frozenlist                     1.4.1
fsspec                         2024.6.1
future                         1.0.0
google-auth                    2.3.3
google-auth-oauthlib           0.4.6
grpcio                         1.42.0
idna                           2.10
imageio                        2.34.2
importlib-metadata             4.8.2
importlib-resources            5.4.0
ipykernel                      6.5.1
ipython                        7.29.0
ipython-genutils               0.2.0
ipywidgets                     7.6.5
jedi                           0.18.1
Jinja2                         3.0.3
json5                          0.9.6
jsonschema                     4.2.1
jupyter-client                 7.1.0
jupyter-core                   4.9.1
jupyter-server                 1.12.0
jupyterlab                     3.2.4
jupyterlab-language-pack-zh-CN 3.2.post2
jupyterlab-pygments            0.1.2
jupyterlab-server              2.8.2
jupyterlab-widgets             1.0.2
kiwisolver                     1.3.2
lazy-loader                    0.4
lightning-utilities            0.11.5
lmdb                           1.5.1
Markdown                       3.3.6
MarkupSafe                     2.0.1
matplotlib                     3.5.0
matplotlib-inline              0.1.3
mistune                        0.8.4
multidict                      6.0.5
nbclassic                      0.3.4
nbclient                       0.5.9
nbconvert                      6.3.0
nbformat                       5.1.3
nest-asyncio                   1.5.1
networkx                       3.1
notebook                       6.4.6
numpy                          1.21.4
oauthlib                       3.1.1
opencv-python                  4.10.0.84
packaging                      21.3
pandocfilters                  1.5.0
parso                          0.8.2
pexpect                        4.8.0
pickleshare                    0.7.5
Pillow                         9.5.0
pip                            21.1.3
prometheus-client              0.12.0
prompt-toolkit                 3.0.22
protobuf                       3.19.1
ptyprocess                     0.7.0
pyasn1                         0.4.8
pyasn1-modules                 0.2.8
pycosat                        0.6.3
pycparser                      2.20
pyDeprecate                    0.3.1
Pygments                       2.10.0
pyOpenSSL                      20.0.1
pyparsing                      3.0.6
pyrsistent                     0.18.0
PySocks                        1.7.1
python-dateutil                2.8.2
pytorch-lightning              1.5.5
pytz                           2021.3
PyWavelets                     1.4.1
PyYAML                         6.0.1
pyzmq                          22.3.0
requests                       2.25.1
requests-oauthlib              1.3.0
rsa                            4.8
ruamel-yaml-conda              0.15.100
scikit-image                   0.21.0
scipy                          1.10.1
Send2Trash                     1.8.0
setuptools                     52.0.0.post20210125
setuptools-scm                 6.3.2
six                            1.16.0
sniffio                        1.2.0
supervisor                     4.2.2
tensorboard                    2.7.0
tensorboard-data-server        0.6.1
tensorboard-plugin-wit         1.8.0
terminado                      0.12.1
testpath                       0.5.0
thop                           0.1.1.post2209072238
tifffile                       2023.7.10
tomli                          1.2.2
torch                          1.10.0+cu113
torchmetrics                   1.4.0.post0
torchvision                    0.11.1+cu113
tornado                        6.1
tqdm                           4.61.2
traitlets                      5.1.1
typing-extensions              4.0.0
urllib3                        1.26.6
wcwidth                        0.2.5
webencodings                   0.5.1
websocket-client               1.2.1
Werkzeug                       2.0.2
wheel                          0.36.2
widgetsnbextension             3.5.2
yarl                           1.9.4
zipp                           3.6.0

-------
Since this work has been done for a long time, I can't find the original version of the pre-trained weights, I will re-train the model and upload the checkpoint later.

**Comment by Morayabb** [comment]
Thank you very much for your reply. I will try running it again after updating the environment.

**Comment by Morayabb** [comment]
After the environment is set up, the problem of loss=nan still occurs, and when running the code, there is also a problem: **Runtime Warning: invalid value counted in cast np.transpose(quantize(tensor, rgb_range).numpy(), (1, 2, 0)).astype(np.uint8)** 
There may have been an issue during the conversion process when searching for information, and I am using the DIV2K_train_SR dataset.
If you have time and have encountered it before, could you please help solve it, thank you very much.

**Comment by Morayabb** [comment]
After setting up the environment, the problem of loss=nan still occurred. After checking the code, it was found that the tensor2uint8() function had a problem. After modification, nan appeared in the value of sr again. We hope you can help solve this problem when you have time. Thank you very much!
所有改变都在multiscale_sr_model.py.
The code used to check the nan value is:
def check_for_nansr(tensor):
    if torch.isnan(tensor).any():
        print("NaN detected in tensor sr!")
A modification point in the code：
    def training_step(self, batch, *args):
        lr, hr = batch
        out_size = hr.shape[2:]
        sr = self.forward(lr, out_size)
        check_for_nansr(sr)
        loss = self.loss_fn(sr, hr)
        return loss
![image](https://github.com/user-attachments/assets/845c31c4-04ca-44b5-b1ee-0464453d1f8c)
